JSON reporter doesn't work (no output)

The issue occur to me on mocha v3.0.2 and node v5.8.0 on Mac OS X. Here's my test file

``` javascript
'use strict';
var app = require('./app');

describe('Converting strings.', function() {

  it('Can convert a string with one space', function() {
    var result = app.convertString('hello world');
    result.should.eql('hello-world');
  });

  it('Replaces two spaces with one dash', function() {
    app.convertString('hello  world').should.eql('hello-world');
  });
});
```

and the code I'm testing

``` javascript
function convertString(str) {
  return str.split(' ').join('-');
}

module.exports = {
  convertString: convertString
};
```

With `json-stream` reporter everything works fine:

```
$ mocha --require should --reporter json-stream test.js
["start",{"total":2}]
["pass",{"title":"Can convert a string with one space","fullTitle":"Converting strings. Can convert a string with one space","duration":2,"currentRetry":0}]
["fail",{"title":"Replaces two spaces with one dash","fullTitle":"Converting strings. Replaces two spaces with one dash","duration":0,"currentRetry":0,"err":"expected 'hello--world' to equal 'hello-world'","stack":"AssertionError: expected 'hello--world' to equal 'hello-world'\n    at Assertion.fail (node_modules/should/cjs/should.js:205:17)\n    at Assertion.Object.defineProperty.value (node_modules/should/cjs/should.js:277:19)\n    at Context.<anonymous> (test.js:12:46)"}]
["end",{"suites":1,"tests":2,"passes":1,"pending":0,"failures":1,"start":"2016-08-12T15:34:45.144Z","end":"2016-08-12T15:34:45.155Z","duration":11}]
```

but with the json reporter nothing comes out:

```
$ mocha --require should --reporter json test.js  # No output!
```

Am I doing something wrong?

--- a/lib/reporters/json.js
+++ b/lib/reporters/json.js
@@ -54,18 +54,7 @@ function JSONReporter (runner) {
 
     runner.testResults = obj;
 
-    var cache = [];
-    process.stdout.write(JSON.stringify(obj, function (key, value) {
-      if (typeof value === 'object' && value !== null) {
-        if (cache.indexOf(value) !== -1) {
-          // Instead of going in a circle, we'll print [object Object]
-          return '' + value;
-        }
-        cache.push(value);
-      }
-
-      return value;
-    }, 2));
+    process.stdout.write(JSON.stringify(obj, null, 2));
   });
 }
 
@@ -78,17 +67,43 @@ function JSONReporter (runner) {
  * @return {Object}
  */
 function clean (test) {
+  var err = test.err || {};
+  if (err instanceof Error) {
+    err = errorJSON(err);
+  }
+
   return {
     title: test.title,
     fullTitle: test.fullTitle(),
     duration: test.duration,
     currentRetry: test.currentRetry(),
-    err: errorJSON(test.err || {})
+    err: cleanCycles(err)
   };
+}
+/**
+ * Replaces any circular references inside `obj` with '[object Object]'
+ *
+ * @api private
+ * @param {Object} obj
+ * @return {Object}
+ */
+function cleanCycles (obj) {
+  var cache = [];
+  return JSON.parse(JSON.stringify(obj, function (key, value) {
+    if (typeof value === 'object' && value !== null) {
+      if (cache.indexOf(value) !== -1) {
+        // Instead of going in a circle, we'll print [object Object]
+        return '' + value;
+      }
+      cache.push(value);
+    }
+
+    return value;
+  }));
 }
 
 /**
- * Transform `error` into a JSON object.
+ * Transform an Error object into a JSON object.
  *
  * @api private
  * @param {Error} err

--- a/lib/reporters/json.js
+++ b/lib/reporters/json.js
@@ -54,7 +54,18 @@ function JSONReporter (runner) {
 
     runner.testResults = obj;
 
-    process.stdout.write(JSON.stringify(obj, null, 2));
+    var cache = [];
+    process.stdout.write(JSON.stringify(obj, function (key, value) {
+      if (typeof value === 'object' && value !== null) {
+        if (cache.indexOf(value) !== -1) {
+          // Instead of going in a circle, we'll print [object Object]
+          return '' + value;
+        }
+        cache.push(value);
+      }
+
+      return value;
+    }, 2));
   });
 }
 

--- a/lib/reporters/json.js
+++ b/lib/reporters/json.js
@@ -80,6 +80,7 @@ function clean (test) {
     err: cleanCycles(err)
   };
 }
+
 /**
  * Replaces any circular references inside `obj` with '[object Object]'
  *

--- a/test/reporters/json.spec.js
+++ b/test/reporters/json.spec.js
@@ -72,15 +72,15 @@ describe('json reporter', function () {
     }));
 
     runner.run(function (failureCount) {
-      failureCount.should.be.exactly(1);
-      runner.should.have.property('testResults');
-      runner.testResults.should.have.property('failures');
-      runner.testResults.failures.should.be.an.instanceOf(Array);
-      runner.testResults.failures.should.have.a.lengthOf(1);
+      expect(failureCount).to.equal(1);
+      expect(runner).to.have.property('testResults');
+      expect(runner.testResults).to.have.property('failures');
+      expect(runner.testResults.failures).to.be.an(Array);
+      expect(runner.testResults.failures).to.have.length(1);
 
       var failure = runner.testResults.failures[0];
-      failure.should.have.property('title', testTitle);
-      failure.should.have.properties('err');
+      expect(failure).to.have.property('title', testTitle);
+      expect(failure).to.have.property('err');
 
       done();
     });

--- a/test/reporters/json.spec.js
+++ b/test/reporters/json.spec.js
@@ -58,4 +58,31 @@ describe('json reporter', function () {
       done();
     });
   });
+
+  it('should handle circular objects in errors', function (done) {
+    var testTitle = 'json test 1';
+    function CircleError () {
+      this.message = 'oh shit';
+      this.circular = this;
+    }
+    var error = new CircleError();
+
+    suite.addTest(new Test(testTitle, function (done) {
+      throw error;
+    }));
+
+    runner.run(function (failureCount) {
+      failureCount.should.be.exactly(1);
+      runner.should.have.property('testResults');
+      runner.testResults.should.have.property('failures');
+      runner.testResults.failures.should.be.an.instanceOf(Array);
+      runner.testResults.failures.should.have.a.lengthOf(1);
+
+      var failure = runner.testResults.failures[0];
+      failure.should.have.property('title', testTitle);
+      failure.should.have.properties('err');
+
+      done();
+    });
+  });
 });

